src/lib/parse_format.rs

```diff
@@ -20,45 +10,69 @@ pub enum FormatToken {
     Literal(String),
 }
 
-// This newtype is needed to sidestep StructOpt's Vec behaviour
-#[derive(Debug)]
-pub struct Format(pub Vec<FormatToken>);
+// Get around structopt automatic Vec handling.
+pub(crate) type Format = Vec<FormatToken>;
+
+pub(crate) fn parse_format_string(input: &str) -> Result<Format, String> {
+    let mut tokens = Vec::new();
+    let mut input = input.as_bytes();
 
-pub fn parse_format_string(input: &str) -> Result<Format, String> {
-    match parse_all(input) {
-        Ok(("", v)) => Ok(Format(v)),
-        Err(s) => Err(format!("Format string parse error: {:?}", s)),
-        Ok((s, _)) => Err(format!("Format string parse error near \"{}\"", s)),
+    loop {
+        let (token, rest) = match input {
+            [b'%', b'i', rest @ ..] => (FormatToken::WindowId, rest),
+            [b'%', b'g', rest @ ..] => (FormatToken::Geometry, rest),
+            [b'%', b'w', rest @ ..] => (FormatToken::Width, rest),
+            [b'%', b'h', rest @ ..] => (FormatToken::Height, rest),
+            [b'%', b'x', rest @ ..] => (FormatToken::X, rest),
+            [b'%', b'y', rest @ ..] => (FormatToken::Y, rest),
+            [b'%', b'%', rest @ ..] => (FormatToken::Literal("%".to_owned()), rest),
+            [b'%', c, ..] => break Err(format!("Unknown format '%{}'", char::from(*c))),
```

Comment:
This could be "*c as char" instead of "char::from(*c)"

Reply:
Nice, thanks for showing that.